src/arch/arm/64/kernel/vspace.c

```diff
@@ -233,68 +233,173 @@ BOOT_CODE void map_kernel_frame(paddr_t paddr, pptr_t vaddr, vm_rights_t vm_righ
+     *  memory window, i.e., never modify the result of `ptrFromPAddr()` when the
+     *  physical address might lie within the kernel ELF. (Reads are fine if they
+     *  have been flushed appropriately per B2.11; this should be the case for
+     *  any paging structures).
+     *
+     *  As an addendum, LWN Article "ARM's multiply-mapped memory mess" specifies
+     *  that in ARMv6 specifications this kind of aliasing would be UNPREDICTABLE;
+     *  the current ARMv8 A-Profile specifications no longer contain this wording,
+     *  and instead specify the loss of coherency and other behaviours in B2.11.
+     *
+     */
+
+    const word_t kernel_elf_pd_start = GET_KPT_INDEX(KERNEL_ELF_BASE, KLVL_FRM_ARM_PT_LVL(2));
+    const word_t kernel_elf_pd_end   = GET_KPT_INDEX(KERNEL_ELF_TOP, KLVL_FRM_ARM_PT_LVL(2));
+
+    word_t paddr = KERNEL_ELF_PADDR_BASE;
+    word_t vaddr = KERNEL_ELF_BASE;
+    for (word_t pd_idx = kernel_elf_pd_start; pd_idx < kernel_elf_pd_end; pd_idx++) {
+        armKSGlobalKernelPDs[kernel_mappings_pud_idx][pd_idx]
+            = pte_pte_page_new(
 #ifdef CONFIG_ARM_HYPERVISOR_SUPPORT
-                                                                                                                        0, // XN
+                  /* XN */ 0,
 #else
-                                                                                                                        1, // UXN
+                  /* UXN */ 1,
 #endif
-                                                                                                                        paddr,
-                                                                                                                        0,                        /* global */
-                                                                                                                        1,                        /* access flag */
-                                                                                                                        SMP_TERNARY(SMP_SHARE, 0),        /* Inner-shareable if SMP enabled, otherwise unshared */
-                                                                                                                        0,                        /* VMKernelOnly */
-                                                                                                                        NORMAL
-                                                                                                                    );
+                  /* page_base_address */ paddr,
+                  /* global (nG) */ 0,
+                  /* access flag (AF) */ 1,
+                  /* shareability (SH) */ SMP_TERNARY(SMP_SHARE, 0),
+                  /* AP */ APFromVMRights(VMKernelOnly),
+                  /* AttrIndx */ NORMAL
+              );
+
+        assert(pd_idx == GET_KPT_INDEX(vaddr, KLVL_FRM_ARM_PT_LVL(2)));
```

Comment:
You already checked the first index with above, I would move this one after the increments. Then you always check things before doing something.